### PR TITLE
Adds the ability to delete multiple warnings at once

### LINKED ISSUE #2104: Popups overlay action items when action items are clicked
### Versions

Docker Tag 2022.01.1 Pi-hole v5.8.1 FTL v5.13 Web Interface v5.10.1

### Platform

- OS and version:  Ubuntu 20.04

- Platform:  Docker

### Expected behavior

When there are messages in the log, I'd like an easy way to discard them.

### Actual behavior / bug

When an item is discarded, a toast is opened which overlays the action items for log entries. You then have to wait for the toast to disappear, or you hunt after it and hope that you're able to close it before it fades away on its own. This is a terrible game. It's not fun.

### Steps to reproduce

1. Go to Pi-hole diagnosis
2. Click on the trash bin icon
4. See error overlapping other trash bin icons (if not, delete more messages)

## Debug Token

I don't have it.

### PR DESCRIPTION
**Pull request by rdwebdesign**
**By submitting this pull request, I confirm the following:** 

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/AdminLTE/blob/master/CONTRIBUTING.md), as well as this entire template.
- [x] I have made only one major change in my proposed changes.
- [x] I have commented my proposed changes within the code.
- [x] I have tested my proposed changes.
- [x] I am willing to help maintain this change if there are issues with it later.
- [x] I give this submission freely and claim no ownership.
- [x] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [x] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
- [x] I have Signed Off all commits. ([`git commit --signoff`](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff))

---

**What does this PR aim to accomplish?:**

This PR adds new controls to the messages table to allow selection and deletion of multiple warnings at once.

**How does this PR accomplish the above?:**

This PR adds Datatables extensions, buttons and CSS.
Fix: update icon count after delete a warning. Also remove the triangle icon after delete last warning.

**What documentation changes (if any) are needed to support this PR?:**
none.

### CODE REVIEW

**Comment by DL6ER** [comment]
Something isn't right about the messages when mass-deleting:

Screenshot from 2022-02-26 10-13-47

The blue boxes never turn to green. The deletion still worked for all selected messages. It would be cleaner if there is only one popup saying that `message(s) # 6491,6492,6493,6494,6495` have been deleted instead of five individual boxes.

PS: Purely cosmetic but maybe we should disable the buttons when there are no issues in the table:

Screenshot from 2022-02-26 10-12-02

Also I'm wondering about the position of these buttons. Mostly thinking about mobile displays that will not show the entire page at once: should they be at the bottom of the table? You typically scroll down on a page when doing the selections. Or we might want to have them both at the top and the bottom.

**Comment by DL6ER** [comment]
More thoughts about usability:
1. Replace the "inversion" by a "select all" button. I don't think there is really any need for inversion and it takes one more step to realize that an inversion of an empty selection is effectively the same as 
2. If we have them both at the top and the bottom, it might be cleaner to use icons instead of text: and:
   1. Use the trash icon for delete (proposal: show the icon only if at least one item is selected),
   3. Use another icon for the proposed Select All button. Here are some icons I consider useful in this context:  [check](https://fontawesome.com/v5/icons/check?s=solid), [check-square](https://fontawesome.com/v5/icons/check-square?s=solid), [bars](https://fontawesome.com/v5/icons/bars?s=solid), [ballot](https://fontawesome.com/v5/icons/ballot?s=solid), [check-double](https://fontawesome.com/v5/icons/check-double?s=solid)

**Comment by rdwebdesign** [comment]
- The texts were replaced by icons, on the buttons.
image
 - I included buttons at the end of the table, too.
- Now, the buttons only show up if there are messages on the list.

File: scripts/pi-hole/js/messages.js [hunk 1]
```diff
@@ -211,40 +262,63 @@ $(function () {
...
+    if (Object.hasOwnProperty.call(ids, id)) {
+      // Exploit prevention: Return early for non-numeric IDs
+      if (typeof ids[id] !== "number") return;
+      trs.push(ids[id]);
+    }
+  }
 
   utils.disableAll();
-  utils.showAlert("info", "", "Deleting message with ID " + parseInt(id, 10), "...");
+  var idstring = JSON.stringify(ids);
```
Comment (rdwebdesign) [inline scripts/pi-hole/js/messages.js]: I noticed you are creating a filtered array (`trs`) with only numeric values from `ids`, but on line 289 you are using the unfiltered variable. I think you should use:

```diff
-  var idstring = JSON.stringify(ids);
+  var idstring = JSON.stringify(trs);
```

Comment (DL6ER) [reply scripts/pi-hole/js/messages.js]: It doesn't matter at all. If there is a single non-numeric entry in `ids`, the entire function will return early and never reach this point.

**Review by yubiuser — COMMENTED** [review]
1) Form a UX perspective, I would expect a "deselect all" happening once all items have been selected an I click on the "select all" button again
2) Although I have enough space, the new column adds a horizontal scrolling bar which does not look very nice.

Bildschirmfoto zu 2022-02-28 09-40-07

**Comment by rdwebdesign** [comment]
- buttons appear only when needed:
  - "Select All" when nothing is selected;
  - "Delete Selected" only if something is selected;
  - "Select All" (with different icon) when one or more items are selected;
  - "Deselect All" when whole page is selected;
- show buttons at the top and bottom of the table;
- batch deletion accepts array (multiple notifications fixed);
- fixed horizontal bar for long warning messages.

**Review by DL6ER — DISMISSED** [review]

File: scripts/pi-hole/js/messages.js [hunk 2]
```diff
@@ -168,16 +182,63 @@ $(function () {
...
+        text: '<span class="far fa-square"></span>',
+        titleAttr: "Select All",
+        className: "btn-sm datatable-bt selectAll",
+        action: function () {
+          table.rows({ page: "current" }).select();
+        },
+      },
+      {
+        text: '<span class="far fa-minus-square"></span>',
+        titleAttr: "Select More",
```
Comment (yubiuser) [inline scripts/pi-hole/js/messages.js]:
```suggestion
        titleAttr: "Select All",
```

If some items are pre-selected, this basically selects them all

File: scripts/pi-hole/js/messages.js [hunk 3]
```diff
@@ -168,16 +182,63 @@ $(function () {
...
+      {
+        text: '<span class="far fa-square"></span>',
+        titleAttr: "Select All",
+        className: "btn-sm datatable-bt selectAll",
+        action: function () {
+          table.rows({ page: "current" }).select();
+        },
+      },
+      {
+        text: '<span class="far fa-minus-square"></span>',
```
Comment (yubiuser) [inline scripts/pi-hole/js/messages.js]:
```suggestion
        text: '<span class="far fa-plus-square"></span>',
```

Minus could be interpreted as "Deselect all" but a click on the button adds all items

**Review by yubiuser — APPROVED** [review]

**Comment by yubiuser** [comment]
Feature request: port this to all other tables 😁

**Comment by rdwebdesign** [comment]
> Feature request: port this to all other tables 😁

We receive a high number of requests, but as a free open-source project run by volunteers we have very limited resources.
Please, we ask that you open your Feature Requests at our [Discourse Forum](https://discourse.pi-hole.net/c/feature-requests/8).

Thank you for your understanding.

:rofl: :rofl: :rofl:

**Comment by yubiuser** [comment]
https://discourse.pi-hole.net/t/bulk-deletion-of-blacklist-whitelist-entries/32925

**Pull request merged**